Make main composer.json a regular, versioned, manually edited file and decommission top level group variables interpolation

rija opened the pull request:
# Pull request for issue: 

gigascience/security#4, gigascience/security#14

This is a pull request for the following functionalities:

* Replace templated composer.json with the real one at the top level of repo
* Remove usage of group variables in configuraiton 

## How to test?

Run all the tests locally and on CI pipeline

## How have functionalities been implemented?

gigascience/security#4:

Got rid of the template for `composer.json`, created new composer.json at top level, making sure 
all versions are specified, especially for Yii dependencies.
The presence of a real `composer.json` file in top-level directory will allow **Dependabot** to create pull request for vulnerability issues.

gigascience/security#14:

Removed from the multiple configuration bash scripts the code that retrieve variables from top-level directory (`gigascience`) (as represented by the value of `GROUP_VARIABLES_URL`).
(there is no longer variables to retrieve anyway now).
Also remove mention of the variable `GROUP_VARIABLES_URL` from the Gitlab configuration files and from example configuration files.

This is all so that we can assign Gitlab members to sub-groups based on whether they need access to production pipeline or not.

## Any issues with implementation?

n/a

## Any changes to automated tests?

n/a

## Any changes to documentation?

Removed mention of usage of group variables

## Any technical debt repayment?

n/a

## Any improvements to CI/CD pipeline?

n/a

rija committed:
feat: make main composer.json a regular, versioned, manually edited file

rija committed:
feat: remove usage of GROUP_VARIABLES_URL in codebase

kencho51 commented:
Hi @rija,

When I executed `./up.sh` and reached the step `docker-compose exec -T application composer install`, I would get this warning message:

```
docker-compose exec -T application composer install
Composer could not detect the root package (gigascience/gigadb-website) version, defaulting to '1.0.0'. See https://getcomposer.org/root-version
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
Nothing to install, update or remove
Package container-interop/container-interop is abandoned, you should avoid using it. Use psr/container instead.
Package facebook/webdriver is abandoned, you should avoid using it. Use php-webdriver/webdriver instead.
Package hoa/consistency is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/console is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/event is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/exception is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/file is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/iterator is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/protocol is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/stream is abandoned, you should avoid using it. No replacement was suggested.
Package hoa/ustring is abandoned, you should avoid using it. No replacement was suggested.
Package padraic/phar-updater is abandoned, you should avoid using it. No replacement was suggested.
Package phpunit/dbunit is abandoned, you should avoid using it. No replacement was suggested.
Package phpunit/php-token-stream is abandoned, you should avoid using it. No replacement was suggested.
Package phpunit/phpunit-mock-objects is abandoned, you should avoid using it. No replacement was suggested.
Package sebastian/finder-facade is abandoned, you should avoid using it. No replacement was suggested.
Package zendframework/zend-cache is abandoned, you should avoid using it. Use laminas/laminas-cache instead.
Package zendframework/zend-config is abandoned, you should avoid using it. Use laminas/laminas-config instead.
Package zendframework/zend-eventmanager is abandoned, you should avoid using it. Use laminas/laminas-eventmanager instead.
Package zendframework/zend-filter is abandoned, you should avoid using it. Use laminas/laminas-filter instead.
Package zendframework/zend-hydrator is abandoned, you should avoid using it. Use laminas/laminas-hydrator instead.
Package zendframework/zend-i18n is abandoned, you should avoid using it. Use laminas/laminas-i18n instead.
Package zendframework/zend-json is abandoned, you should avoid using it. Use laminas/laminas-json instead.
Package zendframework/zend-serializer is abandoned, you should avoid using it. Use laminas/laminas-serializer instead.
Package zendframework/zend-servicemanager is abandoned, you should avoid using it. Use laminas/laminas-servicemanager instead.
Package zendframework/zend-stdlib is abandoned, you should avoid using it. Use laminas/laminas-stdlib instead.
Generating autoload files
38 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```


It seems to me that the `composer.lock` is not updated, then I executed `% docker-compose exec -T application composer update`, the `composer.lock` is updated. 

```
% docker-compose exec -T application composer update 
Composer could not detect the root package (gigascience/gigadb-website) version, defaulting to '1.0.0'. See https://getcomposer.org/root-version
Loading composer repositories with package information
Updating dependencies
Lock file operations: 5 installs, 67 updates, 2 removals
...
Generating autoload files
45 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
Found 11 security vulnerability advisories affecting 6 packages.
Run "composer audit" for a full list of advisories.

```

**question:** Should the `composer.lock` be updated in this PR as the composer.json is updated?

rija committed:
fix: update the composer.lock to match composer.json

rija committed:
fix: add specific version of AWS SDK in composer.json due to error

after updating composer.lock, transitive dependency updated aws sdk to a version not supported by PHP 7.4,
causing test failure.
So have to explicitely declare the version of AWS SDK we use in the develop branch

kencho51 commented:
Hi @rija,

Thanks for the update.

But now I got the following warnings:
```
.......................................WWWWWW.   63 / 63 (100%)

Time: 1.2 seconds, Memory: 20.00MB

There were 6 warnings:

1) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "valid email and no details" (array('foo@hotmail.com'), 1, array('foo@hotmail.com', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

2) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "valid email and name" (array('foo@hotmail.com', 'Foo', 'Bar'), 1, array('foo@hotmail.com', array('Foo', 'Bar'), 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

3) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "obvious invalid email" (array('foobar'), 0, array('foobar', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

4) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "subtle invalid email" (array('"much.more unusual"@example.com'), 0, array('"much.more unusual"@example.com', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

5) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "Email domain doesn't exist" (array('james@speed.bump'), 0, array('james@speed.bump', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

6) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "Internationalized email address" (array('ουτοπία@δπθ.gr'), 1, array('xn--kxae4bafwg@xn--pxaix.gr', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

WARNINGS!
Tests: 63, Assertions: 150, Warnings: 6.
```

kencho51 reviewed (approved):
Hi @rija,

**praise:** The composer.json is being versioned and github dependabot can use it to detect the vulnerabilities.
**praise:** The extra `GROUP_VARIABLES_URL` has been removed, and the related generate config bash scripts have been updated accordingly.
**note (non-blocking):** Got the following warnings when `./tests/unit_runner`:

```
.......................................WWWWWW.   63 / 63 (100%)

Time: 1.2 seconds, Memory: 20.00MB


There were 6 warnings:

1) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "valid email and no details" (array('foo@hotmail.com'), 1, array('foo@hotmail.com', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

2) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "valid email and name" (array('foo@hotmail.com', 'Foo', 'Bar'), 1, array('foo@hotmail.com', array('Foo', 'Bar'), 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

3) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "obvious invalid email" (array('foobar'), 0, array('foobar', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

4) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "subtle invalid email" (array('"much.more unusual"@example.com'), 0, array('"much.more unusual"@example.com', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

5) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "Email domain doesn't exist" (array('james@speed.bump'), 0, array('james@speed.bump', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

6) NewsletterServiceTest::testItShouldUnsubscribeUser with data set "Internationalized email address" (array('ουτοπία@δπθ.gr'), 1, array('xn--kxae4bafwg@xn--pxaix.gr', 'subscribed'))
Static method "subscriberHash" cannot be invoked on mock object

WARNINGS!
Tests: 63, Assertions: 150, Warnings: 6.
```
**praise:** All tests are passing in my dev and also the CI/CD pipeline.

No blockings from me, happy to approve.

pli888 reviewed (approved):
**praise**: I can see that the `composer.json.dist` has been removed and there is now a `composer.json` in root directory of repo.
**praise**: All tests are running in my dev and staging environments. Code looks fine too.

Happy to approve.

rija committed:
docs: add entry to CHANGELOG.md
